### PR TITLE
[aot_autograd] Handle None in saved tensors for LSTM/RNN backward; enable allow_rnn by default

### LINKED ISSUE #158007: `torch.compile` cannot compile a model with a basic `LSTM`, even on latest main
## Reproduction Script
```python
import torch
import torch.nn as nn

class MyModel(nn.Module):
    def __init__(self, rnn_dims=512):
        super().__init__()
        self.rnn = nn.LSTM(input_size=rnn_dims, hidden_size=rnn_dims, batch_first=True)

    def forward(self, x):
        h0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
        c0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
        out, h1 = self.rnn(x, (h0, c0))
        return h1  # tuple (h_n, c_n)

if __name__ == "__main__":
    torch.manual_seed(0)
    model = MyModel().eval().to("cuda")
    input_tensor = torch.rand(1, 1, 512, dtype=torch.float32).to("cuda")

    try:
        with torch.no_grad():
            eager_output = model(input_tensor)
        print("Eager output:", eager_output[0].shape, eager_output[1].shape)
    except Exception as e:
        print(e)

    try:
        compiled_model = torch.compile(model, fullgraph=True, backend="inductor")
        with torch.no_grad():
            compiled_output = compiled_model(input_tensor)
        print("Compiled output:", compiled_output[0].shape, compiled_output[1].shape)
    except Exception as e:
        print(e)
```

## Output on Stable (2.7.1) and Latest `main` (2.9.0a0+gitfcc682b)

<details> <summary>Output (Stable)</summary>

```
Eager output: torch.Size([1, 1, 512]) torch.Size([1, 1, 512])
TorchDynamo purposely graph breaks on RNN, GRU, LSTMs

from user code:
   File "for_test.py", line 10, in forward
    h0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
```

</details> 

<details> <summary>Output (main branch)</summary>

```
Eager output: torch.Size([1, 1, 512]) torch.Size([1, 1, 512])
Attempted to wrap RNN, GRU, or LSTM
  Explanation: Dynamo does not support RNN, GRU, or LSTM.
  Hint: It may be possible to write Dynamo tracing rules for this code. Please report an issue to PyTorch if you encounter this graph break often and it is causing performance issues.

  Developer debug context: LSTM(512, 512, batch_first=True)

from user code:
   File "test_coverage/for_test.py", line 10, in forward
    h0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
```

</details>

## What’s Confusing?
+ This model runs just fine under eager execution.

+ It’s a very basic LSTM-based model, and doesn’t use anything exotic or fused.

+ Yet torch.compile breaks, even on the latest main (torch==2.9.0a0+gitfcc682b).

+ I understand some ops are not supported yet — but shouldn't a warning or fallback happen, instead of an outright failure? Or is there some plan to support nn.LSTM and friends via fake tensor rules in the future?

### Versions

Both `torch==2.9.0a0+gitfcc682b` and `torch==2.7.1a0+gitfcc682b` have the same problem.

```
PyTorch version: 2.7.1a0+gite2d141d
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
```

Thanks for the great work on `torch.compile` — I appreciate any clarification on this!

### PR DESCRIPTION
**Pull request by XAheli**
Fixes #158007

`torch.compile` now works with `nn.LSTM`, `nn.GRU`, and `nn.RNN` modules on both CPU (MKLDNN) and CUDA paths. This is a correctness fix i.e. users who called `torch.compile(model_with_lstm)` got a crash; they now get a working compiled model. Eager behavior is unaffected.

> **Note:** compilation does not currently speed up LSTM-dominated models. On CUDA, cuDNN's fused kernel is already highly optimized, and the compiled path decomposes it into cell-level ops. The value of this change is removing the graph break so that (a) `torch.compile` doesn't crash, and (b) surrounding non-RNN ops in the model benefit from compilation.

## Root causes

Three independent blockers prevented LSTM from compiling:

**A. Dynamo graph-breaks on RNN modules.** `allow_rnn` defaulted to `False`, causing an immediate graph break on any RNN module. Fix: flip to `True`.

**B. Dynamo graph-break from `flatten_parameters()`.** Models calling `flatten_parameters()` in forward hit `torch.cuda.device_of` on the Dynamo skip-list. Fix: no-op during `torch.compiler.is_compiling()`, same pattern as [`transformer.py:476`](https://github.com/pytorch/pytorch/blob/b2329090622d/torch/nn/modules/transformer.py#L476).

**C. MKLDNN workspace null during compiled backward.** Two bugs:

1. The decomposition [`mkldnn_one_layer_lstm`](https://github.com/pytorch/pytorch/blob/b2329090622d/torch/_decomp/decompositions.py#L3936) hardcoded `train=False` at [line 3958](https://github.com/pytorch/pytorch/blob/b2329090622d/torch/_decomp/decompositions.py#L3958). This decomposition is selected when `input.requires_grad=False` (typical for user data), but model parameters still require grad and need backward. The hardcoded value told the C++ kernel "no backward needed" when backward was actually needed for parameter gradients. Fix: `train = w0.requires_grad or w1.requires_grad`.

2. The C++ kernel [`mkldnn_rnn_layer`](https://github.com/pytorch/pytorch/blob/b2329090622d/aten/src/ATen/native/mkldnn/RNN.cpp#L213) gated workspace allocation on `at::GradMode::is_enabled()` at [line 273](https://github.com/pytorch/pytorch/blob/b2329090622d/aten/src/ATen/native/mkldnn/RNN.cpp#L273) instead of the `train` parameter. AOTAutograd's compiled forward runs under `no_grad`, so GradMode was always false in the compiled context. The `train` parameter is the correct semantic indicator (confirmed via [`derivatives.yaml`](https://github.com/pytorch/pytorch/blob/b2329090622d/tools/autograd/derivatives.yaml): backward always consumes `result3` workspace). Fix: `if (train || at::GradMode::is_enabled())`. The inference primitive is preserved for genuine inference (`train=False`, no GradMode).

## Test plan

- Verified exact reproducer from #158007 compiles and produces correct output on CPU and CUDA
- Tested LSTM, GRU, RNN forward + backward with numeric equality checks against eager (bit-identical on CPU MKLDNN path)
- Tested bidirectional, multi-layer, with/without dropout
- Tested inference (`eval` + `no_grad`) and training modes
- Verified zero graph breaks for standalone LSTM (`fullgraph=True`)
- Verified eager performance/behavior is unaffected (C++ gate evaluates identically in all standard eager scenarios)
- Existing tests pass: `test_aot_autograd` (7/7), `test_cpu_repro` LSTM tests

Co-authored with Opus 4.6

cc [@mikaylagawarecki](https://github.com/mikaylagawarecki) [@jgong5](https://github.com/jgong5) [@mingfeima](https://github.com/mingfeima) [@XiaobingSuper](https://github.com/XiaobingSuper) [@sanchitintel](https://github.com/sanchitintel) [@ashokei](https://github.com/ashokei) [@jingxu10](https://github.com/jingxu10) [@aditew01](https://github.com/aditew01) [@voznesenskym](https://github.com/voznesenskym) [@penguinwu](https://github.com/penguinwu) [@EikanWang](https://github.com/EikanWang) [@Guobing-Chen](https://github.com/Guobing-Chen) [@zhuhaozhe](https://github.com/zhuhaozhe) [@blzheng](https://github.com/blzheng) [@wenzhe-nrv](https://github.com/wenzhe-nrv) [@jiayisunx](https://github.com/jiayisunx) [@ipiszy](https://github.com/ipiszy) [@kadeng](https://github.com/kadeng) [@muchulee8](https://github.com/muchulee8) [@amjames](https://github.com/amjames) [@chauhang](https://github.com/chauhang) [@aakhundov](https://github.com/aakhundov) [@coconutruben](https://github.com/coconutruben) [@jataylo](https://github.com/jataylo) [@azahed98](https://github.com/azahed98) [@jansel](https://github.com/jansel) [@aorenste](https://github.com/aorenste) [@bdhirsh](https://github.com/bdhirsh)

### CODE REVIEW

**Comment by XAheli** [comment]
FYI seems like the failure is related to #174929

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 1]
```diff
@@ -3199,6 +3229,17 @@ def _codegen_compiled_forward(
     )
 
 
+def _reinsert_saved_nones(ctx: Any, saved: tuple[Any, ...]) -> tuple[Any, ...]:
+    vc_nones = getattr(ctx, "_none_positions_vc", [])
+    no_vc_nones = getattr(ctx, "_none_positions_no_vc", [])
+    if not vc_nones and not no_vc_nones:
+        return saved
+    result = list(saved)
+    for pos in sorted(vc_nones + no_vc_nones):
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: (Claude Review) [blocker] Positional offset bug. `ctx._none_positions_no_vc` are positions within the *no_vc sub-list*, not the combined `_saved` tuple. At the call site, `_saved = (*_ctx_.saved_tensors, *_ctx_._tensors_no_vc_check)` — the no_vc portion starts at offset `num_vc_check` (the original vc length, including any vc Nones). Re-inserting them at their raw sub-list index puts the None inside the vc portion of the combined tuple, shifting every subsequent tensor and silently corrupting the positional contract with the backward graph.

Example 1: `tensors_saved_with_vc_check = [V0, V1, V2]` (no Nones), `tensors_saved_no_vc_check = [T0, None, T2]`.
- `_none_positions_vc=[]`, `_none_positions_no_vc=[1]`
- `_saved = (V0, V1, V2, T0, T2)`
- correct reconstruction: `(V0, V1, V2, T0, None, T2)` — None at combined index 4
- actual result: `result.insert(1, None)` → `(V0, None, V1, V2, T0, T2)` — None at index 1

Example 2: `tensors_saved_with_vc_check = [V0, None, V2]`, `tensors_saved_no_vc_check = [T0, None, T2]`.
- correct: `(V0, None, V2, T0, None, T2)` — Nones at 1 and 4
- actual: `sorted([1, 1])` → inserts at 1 twice → `(V0, None, None, V2, T0, T2)` — Nones clustered at 1 and 2

Fix: offset the no_vc positions by `num_vc_check` (the original vc length, before filtering) before sorting:
```python
num_vc_check_orig = len(ctx.saved_tensors) + len(vc_nones)
positions = sorted(vc_nones + [num_vc_check_orig + p for p in no_vc_nones])
for pos in positions:
    result.insert(pos, None)
```
This also requires either recording `num_vc_check` on `ctx` or computing it as above (which works since `len(ctx.saved_tensors)` equals the filtered vc count). See the design note in the review body — the `save_for_backward`-supports-None route would obviate this helper entirely.

**Review by aorenste — CHANGES_REQUESTED** [review]

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 1 above)
Comment (XAheli) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: switched to passing None through `save_for_backward`. Updated in the latest commit

**Comment by XAheli** [comment]
Hi @aorenste @bdhirsh Could you PTAL at the latest changes! Thanks :)

**Comment by XAheli** [comment]
FYI traced the CI failures e2e on H200 with source-built torch/torchvision/torchaudio with claude

- `doctr_reco_predictor` (graph_breaks: 1→3) This is not caused by LSTM tracing at all. With `allow_rnn=False` the LSTM module boundary break happened first and the rest of the model fell back to eager, hiding downstream breaks. With `allow_rnn=True` the LSTM now compiles successfully, more of the model gets traced and doctr's own `@torch.compiler.disable` decorator on its postprocessor (`doctr/models/recognition/crnn/pytorch.py:226`) becomes visible. So more of the model compiles, the higher break count just reflects more code being traced.

- `tts_angular` (graph_breaks: 2→3) This is a graph break from `flatten_parameters()`. The tts_angular model has 3 stacked `LSTMWithProjection` layers that each call `self.lstm.flatten_parameters()` in their forward. `flatten_parameters()` at `rnn.py:275` uses `torch.cuda.device_of(first_fw)`, which hits the Dynamo MOD_SKIPLIST (`torch.cuda` module) and breaks the graph. With `allow_rnn=False` this was never reached because Dynamo broke at the module boundary first.

- `flatten_parameters()` is a purely imperative cuDNN weight layout optimization. It rearranges parameter data pointers for cuDNN's packed format. During Dynamo tracing this is unnecessary since the traced graph captures mathematical ops, not weight layout. In the recent commit I made `flatten_parameters()` a no-op during `torch.compiler.is_compiling()`, same pattern used in `transformer.py:476` and `module.py:1880`.

The other CI failures are all unrelated to this PR.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Comment by XAheli** [comment]
Hi @jansel could you PTAL I've addressed the reviews and added test cases

**Comment by jansel** [comment]
@claude review this again

**Comment by jithunnair-amd** [comment]
This PR's merge base is too old, so it's looking for old runner labels.

@pytorchbot rebase
